### ISSUE TITLE
JNDI incompletly populated for EJBs in WAR

### ISSUE THREAD

**Issue by matthiasblaesing** [payara/Payara#401]
I was faced with this problem last week and had now some time to do some reading and experimenting.

For the following I reference the EJB 3.2 contract (from ejb-3_2-cor-fr-spec.pdf, JSR 345: Enterprise JavaBeans TM ,Version 3.2 EJB Core Contracts and Requirements).

For reference I pushed a sample application to:

https://github.com/matthiasblaesing/PayaraWarJNDI

A build results in a Servlet and an EJB (stateless session bean, eu.doppel_helix.test.jnditest.ejb.Sample). The EJB is meant as a sample of the problem. The servlet outputs the JNDI.

From the chapter ```4.4 Access to the Global JNDI Namespace``` the above sample bean should register:

* [4.4.1] java:global/jnditest-1.0-SNAPSHOT/Sample
* [4.4.1] java:global/jnditest-1.0-SNAPSHOT/Sample!eu.doppel_helix.test.jnditest.ejb.Sample
* [4.4.1.1] java:app/jnditest-1.0-SNAPSHOT/Sample
* [4.4.1.1] java:app/jnditest-1.0-SNAPSHOT/Sample!eu.doppel_helix.test.jnditest.ejb.Sample
* [4.4.1.2] java:module/Sample
* [4.4.1.2] java:module/Sample!eu.doppel_helix.test.jnditest.ejb.Sample

The matching excample would be (4.4.2.1, page 79):

```
package com.acme;
@Stateless
public class FooBean implements Foo { ... }
```

[...]

```
If FooBean is packaged in a stand-alone fooweb.war file, without the use of any deployment
descriptors, the resulting global JNDI name entries are:
java:global/fooweb/FooBean
java:global/fooweb/FooBean!com.acme.Foo
java:app/fooweb/FooBean
java:app/fooweb/FooBean!com.acme.Foo
java:module/FooBean
java:module/FooBean!com.acme.Foo
```

Replacing the placeholders:

* fooweb => jnditest-1.0-SNAPSHOT
* FooBean => Sample
* com.acme.Foo => eu.doppel_helix.test.jnditest.ejb.Sample

transforms my expectation into the sample.

The sample programm outputs:

```
Sample woks: Hallo Welt

========= java:comp =========
env => com.sun.enterprise.naming.impl.JavaURLContext
    eu.doppel_helix.test.jnditest.JNDILister => com.sun.enterprise.naming.impl.JavaURLContext
        sample => eu.doppel_helix.test.jnditest.ejb.__EJB31_Generated__Sample__Intf____Bean__
========= java:global =========
jnditest-1.0-SNAPSHOT => com.sun.enterprise.naming.impl.TransientContext
    Sample => com.sun.ejb.containers.JavaGlobalJndiNamingObjectProxy
    Sample!eu.doppel_helix.test.jnditest.ejb.Sample => com.sun.ejb.containers.JavaGlobalJndiNamingObjectProxy
mejb => com.sun.enterprise.naming.impl.TransientContext
    MEJBBean => org.glassfish.kernel.javaee.MEJBNamingObjectProxy
    MEJBBean!org.glassfish.admin.mejb.MEJBHome => org.glassfish.kernel.javaee.MEJBNamingObjectProxy
========= java:app =========
env => com.sun.enterprise.naming.impl.JavaURLContext
========= java:module =========
env => com.sun.enterprise.naming.impl.JavaURLContext
    eu.doppel_helix.test.jnditest.JNDILister => com.sun.enterprise.naming.impl.JavaURLContext
        sample => eu.doppel_helix.test.jnditest.ejb.__EJB31_Generated__Sample__Intf____Bean__
```

The app and module namespaces are enumerated, but are empty.

**Comment by smillidge** [comment]
I've added a pull request to your sample. If you add direct lookup code to the sample
```java
                out.append("=============  Direct look up of bean using the JNDI names ==================\n");
                out.append("java:app/jnditest-1.0-SNAPSHOT/Sample = " + ic.lookup("java:app/jnditest-1.0-SNAPSHOT/Sample").toString() + "\n");
                out.append("java:app/jnditest-1.0-SNAPSHOT/Sample!eu.doppel_helix.test.jnditest.ejb.Sample = " + ic.lookup("java:app/jnditest-1.0-SNAPSHOT/Sample!eu.doppel_helix.test.jnditest.ejb.Sample").toString() + "\n");
                out.append("java:module/Sample = " + ic.lookup("java:module/Sample").toString() + "\n");
                out.append("java:module/Sample = " + ic.lookup("java:module/Sample!eu.doppel_helix.test.jnditest.ejb.Sample").toString() + "\n");
```

The names are resolvable as expected.
=============  Direct look up of bean using the JNDI names ==================
java:app/jnditest-1.0-SNAPSHOT/Sample = eu.doppel_helix.test.jnditest.ejb.Sample@58fdacd5
java:app/jnditest-1.0-SNAPSHOT/Sample!eu.doppel_helix.test.jnditest.ejb.Sample = eu.doppel_helix.test.jnditest.ejb.Sample@58fdacd5
java:module/Sample = eu.doppel_helix.test.jnditest.ejb.Sample@58fdacd5
java:module/Sample = eu.doppel_helix.test.jnditest.ejb.Sample@58fdacd5

My understanding is the JavaURLContext in the JNDI tree at each of the namespaces is performing a dynamic lookup when requested.

**Comment by matthiasblaesing** [comment]
Thank you for taking a look at this. I did not take into account, that JNDI is kind of special and was under the wrong impression that I should be able to enumerate the entries. But rereading that section this is not that clear ...

As I can't reproduce the exact circumstances of my problem last week (I now tested also with different approaches in entity listeners and everything work) and considering, that I now followed multiple trails I'll close the hunt for me - maybe it was a lack of caffeine. 

Thanks again!
